New post: Deploy an ASP.NET Core Application on Windows Server with AWS Lightsail (#443)
build-on-aws/content: tutorials/practical-cloud-guide/deploy-an-asp-net-core-application-on-windows-server-with-aws-lightsail/index.md

---
title: "Deploy an ASP.NET Core Application on Windows Server with AWS Lightsail"

> description: ""

Reviewer: @spara, can you please add a description for the piece that ties in with the title, but doesn't repeat it?

Author: added

Reviewer: Can you confirm you pushed the commit? It still shows "" for the description on line 3.

| 💻 Code Sample         | Code sample used in tutorial on [GitHub](https://github.com/build-on-aws/practical-cloud-guide-code)                             |
| 📢 Feedback            | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?    |
| ⏰ Last Updated     | 2023-05-31                             |

| ToC |
|-----|

---

## Lightsail Overview

> For this task you will deploy an ASP.NET Core application on IIS in Windows Server. The application is simple web application, but requires installing and configuring IIS in addition to deploying it. Let’s begin with a AWS Lightsail overview to familiarize working with this service.

Reviewer: ```suggestion
For this task you will deploy an ASP.NET Core application on IIS in Windows Server. The application is a simple web application, but requires installing and configuring IIS in addition to deploying the application to IIS. Let’s begin with a AWS Lightsail overview to familiarize working with this service.
```

Reviewer: Can you also add a bit more of an overview of how the solution will use Lighsail, e.g. something like "We will use Lightsail to run a virtual server for us, compile and zip up the application we want to deploy, upload the zip to S3 to allow copying it easily to the server, configure IIS, and then deploy our application"?

> Open the AWS Console in a browser and use the search bar to find AWS Light Sail.

Reviewer: Having gone through the rest of the tutorial, this section isn't clear that you are illustrating how you would spin up a Lightsail instance, but the reader should not follow the steps. Can you please add in a clear statement that this is to illustrate how to spin up a new instance, but that we are going to add in some automation to configure it for us and that the reader should *not* spin this one up?

Author: added statement and anchor ref to go to Module 1 if they want to start the tutorial immediately

![Open AWS Lightsail in console](./images/PCG-1-lightsail.png)

The Lightsail menu displays an option for **Instances**, or Virtual Private Servers (VPS). Choose **Instances**, then choose **Create instance**.

![Create a VPS instance in AWS Lightsail](./images/PCG-2-lightsail.png)

> **Create an instance** has several options. First, choose the **Instance location**, you can leave the default, or choose the closest AWS Region. Second, choose either Linux or Windows Server for the VPS. Third, you can choose a blueprint, which is a VPS with a pre-configured application for Linux instance; or you choose a OS Only VPS. For Windows Server, you can choose a blueprint for SQL Server or choose a version of Windows Server.

Reviewer: Can they choose Linux or Windows? I believe IIS only runs on Windows?

Author: simplified to only Windows

![Create an instance](./images/PCG-3-lightsail.png)

> You can choose the Instance plan. One of the advantages of Lightsail is a fixed monthly cost for a VPS.

Reviewer: ```suggestion
There are multiple instance types to choose from, one of the advantages of Lightsail is a fixed monthly cost for a VPS. For this tutorial, we are going to use the <which size you picked> instance type.
```

![Choose an instance plan](./images/PCG-4-lightsail.png)

This is a brief overview of AWS Lightsail. As we progress through the tasks, we’ll go in depth with Lightsail’s other services.

## Module 1: Create an S3 bucket and upload files

### Overview

> In [DevOps](/concepts/devops-essentials), applications are typically built with Continuous Integration (CI) software. Code is pushed into the CI by developers where it is built and tested and released into cloud storage.

Reviewer: Can you please expand here that we are going to use a powershell script to configure the server for us before showing the steps to allow the reader to understand the why before we get to the what?

Reviewer: Also, after finishing the tutorial, I realized we aren't setting up any kind of automated deployment here, can you please expand the intro to indicate this isn't going to set up a CI server, or update this line to indicate how the powershell script ties in with DevOps and remove the mention of a CI server? The way you phrase it on line 286 is great: 

"The `deploy_iis.ps1` Powershell script shows how you can use familiar scripting tools and commands to automate configuring Windows services such as IIS while interacting with AWS resources. Although a simple example, this shows how to implement a common DevOps workflow of deploying an application from an object store and automating the process with a script."

If you can adapt this to be forward looking - "we are going to use a Powershell script to help us with X, Y, and Z".

Author: addressed

### What you will accomplish

> In this module, the software and deployment script is in a GitHub repository. You will clone the repository to copy the files to your local drive. The next step is to create an S3 bucket to store the files that can be accessed by a Windows Server and deployed.

Reviewer: Please add in a bit of detail about S3 bucket names being globally unique, and state that for this tutorial, you will use the name X.

Author: addressed

### Implementation Instructions

Step 1: Clone the `practical-cloud-guide` repository.

```bash
git clone https://github.com/build-on-aws/practical-cloud-guide-code
```

Step 2: Open the AWS Console and choose Lightsail.

![Open AWS Lightsail](./images/PCG-5-lightsail.png)

Step 3: Create an S3 bucket

Choose **Storage**.

![Choose Storage in the Lightsail menu](./images/lightsail-s3-bucket-1.png)

> In the **Create a new bucket** page choose the **5GB storage plan** and name the bucket `practical-cloud-guide`. Select **Create Bucket**.

Reviewer: Please add a comment here as well that the tutorial uses "practical-cloud-guide" as the bucket name, but they need to pick a different one.

Author: addressed

![Create an S3 bucket](./images/lightsail-s3-bucket-2.png)

You will see a menu page for the `practical-cloud-guide` bucket, choose **Objects**.

![Open the Objects menu](./images/lightsail-s3-bucket-3.png)

The **Object list** displays the objects in the bucket. Choose **Upload** to put the application and deployment file in the bucket.

![Upload files to S3](./images/lightsail-s3-bucket-4.png)

Choose **File**.

![Choose File](./images/lightsail-s3-bucket-5.png)

> Select `app.zip` and `deploy_iis.ps1` from `./practical-cloud-guide-code/run-to-build/windows-app-deploy

Reviewer: I don't see the zip in the sample code repo, and it should not be added. Is there a step here missing on how to compile the application and zip it up?

Author: added instructions for compiling and packaging

### What you will accomplish

Using AWS Lightsail, you will deploy a Windows Server 2022 instance with IIS installed using a Powershell script.

### Implementation instructions

Step 1: Deploy Windows Server 2022

Choose **Create instance**.

![Open the server menu](./images/PCG-6-lightsail.png)

Choose **Microsoft Windows**, then choose **Windows Server 2022**.

![Create a Windowserver instance](./images/PCG-7-lightsail.png)

Choose an instance plan, for this tutorial you can use the smallest plan, but larger plans are more responsive.

![Choose an instance plan](./images/PCG-9-lightsail.png)

Name your instance `Windows_Server_IIS`. Then choose **Create Instance**.

## Module 3: Deploy an ASP.NET Core application

### Overview

In the previous module, you created a Windows 2022 server with Lightsail. The next step is to provision the server with IIS and deploy a web application written in C#.

### What you will accomplish

This module shows how to install and configure IIS in Windows Server 2022 and deploy a ASP.NET Core application from an S3 bucket with a Powershell Script

### Implementation instructions

Step 1: Add your AWS Credentials

> You can use an RDP client or **Use your browser** to log into the Windows server. You will download a deployment Powershell script and the ASP.NET Core application from the S3 bucket. Downloading files from S3 requires your AWS credentials, i.e. the Access Key and the Secret Key.

Reviewer: Can you please add instructions on how to RDP from the browser to the instance?

> Open a Powershell terminal from either the Windows menu bar or by searching for Powershell. The Lightsail Windows Server instances include [AWS Tools for Powershell](https://docs.aws.amazon.com/powershell/), which are a set of Window cmdlets for creating and managing AWS resources.  Set your AWS credentials with Powershell.

Reviewer: ```suggestion
Open a Powershell terminal from either the Windows menu bar or by searching for Powershell. The Lightsail Windows Server instances include [AWS Tools for Powershell](https://docs.aws.amazon.com/powershell/?sc_channel=el&sc_campaign=community-aws&sc_content=deploy-an-asp-net-core-application-on-windows-server-with-aws-lightsail&sc_geo=mult&sc_country=mult&sc_outcome=acq), which are a set of Windows `cmdlets` for creating and managing AWS resources.  Set your AWS credentials with Powershell.
```

> ```powershell

Reviewer: Instead of using hard-coded API credentials, can you please update to use resource permissions that don't require the key/secret please? https://lightsail.aws.amazon.com/ls/docs/en_us/articles/amazon-lightsail-configuring-bucket-resource-access

Author: I haven't covered IAM roles yet, it's in a later post. The idea behind this tutorial is to onboard an IT Pro quickly and successfully. I've changed it so that credentials are added when the Windows VPS is launched and deleted when IIS is configured with a note that it is best practice to not leave credentials on resources that no longer need them.

$YourAccessKey = "your_access_key"
$YourSecretKey = "your)secret_key"
Set-AwsCredential -AccessKey $YourAccessKey -SecretKey $YourSecretKey -StoreAs default
```

Check that your credentials were created. You should see `default` as the ProfileName.

```powershell
Get-AWSCredentials -ListProfileDetail

ProfileName StoreTypeName         ProfileLocation
----------- -------------         ---------------
default     NetSDKCredentialsFile
```

Step 2: Download the IIS install and configuration script from S3.

Create a directory from the root of the C:\ drive and name it `deploy`. Download the `deploy_iis.ps1` script from S3 with the `Copy-S3Object` cmdlet.

```powershell
New-Item -Path C:\deploy -ItemType Directory
cd C:\deploy
$YourBucketName = “practical-cloud-guide”
$YourAppKey = “deploy_iis.ps1”
Copy-S3Object -BucketName $YourBucketName -Key $YourAppKey -LocalFile C:\deploy\$YourAppKey
```

Step 3: Deploy IIS and an ASP.NET Core application

> The deploy_iis.ps1 Powershell script automates the process of installing IIS and its management tools., configuring a new website and deploying a ASP.NET Core Razor application. Let’s break down the script before running it.

Reviewer: ```suggestion
The 1deploy_iis.ps11 Powershell script automates the process of installing IIS and its management tools., configuring a new website and deploying a ASP.NET Core Razor application. Let’s break down the script before running it.
```

The first part of the script installs IIS and the management tools. To host the ASP.NET Core application, IIS requires ASP.NET Core hosting bundle. To learn more about IIS configuration see the IIS documentation. Note that the script sets ProgressPreference to SilentlyContinue to prevent cmdlet outputs from writing to the terminal.

> Set-Variable $global:ProgressPreference SilentlyContinue

Reviewer: ```suggestion
```powershell
Set-Variable $global:ProgressPreference SilentlyContinue
```
```
I don't seem to be able to include the closing triple back-tick here, can you please make sure to add it after the command?

```powershell
# Install IIS
Install-WindowsFeature Web-Server -IncludeManagementTools

# Download and install the ASP.NET Core 6.0 Hosting Bundle
$filein = "https://download.visualstudio.microsoft.com/download/pr/7ab0bc25-5b00-42c3-b7cc-bb8e08f05135/91528a790a28c1f0fe39845decf40e10/dotnet-hosting-6.0.16-win.exe"
Invoke-WebRequest -Uri $filein -OutFile "$(pwd)\dotnet-hosting-6.0.16-win.exe"

Start-Process -FilePath "$(pwd)\dotnet-hosting-6.0.16-win.exe" -Wait -ArgumentList /passive

# Stop and start IIS

> net stop was /y

Reviewer: Is this command correct? Should it be `was` or `w3svc` since I see that is used for the `start` command below?

Author: Windows Activation Service is the part than manages application pools for IIS, w3sc restarts IIS so that the IIS manager is forced to synchs up with the newly installed IIS. Basically a reboot to make sure that the IIS manager is talking to IIS.

net start w3svc
```

> The second part of the script creates an directory for the application. The script downloads app.zip from the S3 bucket created earlier and unzips it on the directory.

Reviewer: ```suggestion
The second part of the script creates an directory for the application. The script downloads `app.zip` from the S3 bucket created earlier and unzips it on the directory.
```

```powershell
# download and unzip the application

> $YourBucketName = "practical-cloud-guide-3"

Reviewer: ```suggestion
$YourBucketName = "practical-cloud-guide"
```

# start new website
Start-WebAppPool -Name $appPoolName
Start-WebSite -Name "DemoSite"

# Open application on Edge
start microsoft-edge:http://localhost:8080/DemoApp
```

Run the script to complete the installation and deployment.

```powershell
C:\deploy\deploy_iis.ps1
```

## Module 4: Clean up

To prevent additional costs, delete the Windows Server 2022. Deleting the S3 bucket is optional. You can keep the S3 bucket to use with other tutorials.

Step 1: Delete Windows Server 2022

Choose **Instances** in the Lightsail Menu and select the three red dots.

![Choose Delete option in the Instance menu](./images/delete-vps-1.png)

Choose **Delete**.

![Choose Delete](./images/delete-vps-2.png)

Choose **Yes, delete**.

![Confirm Delete](./images/delete-vps-3.png)

Step 2: Delete the S3 bucket (Optional)

Choose **Storage** on the Lightsail menu. Select the three vertical dots.

![Choose Delete option in the Storage menu](./images/s3-delete-1.png)

Choose **Delete**.

![Choose Delete](./images/s3-delete-2.png)

Choose **Force Delete** to delete the files and the S3 bucket.

![Confirm Force Delete](./images/s3-delete-3.png)

## What did you accomplish?

> The first cloud resource you created was an S3 bucket to store files that are accessible to cloud services. S3 is an object storage which is different from a file system which supports file read and write. You had to copy files from S3 to the Windows server to work with them.

Reviewer: This is a good place to link to your piece on understanding object/file/block storage.
